Store hints/warnings

rockstardev opened the pull request:
Resolves: #1959

WIP

Illustration of new warnings:

![image](https://user-images.githubusercontent.com/5191402/96217328-f7d5a300-0f47-11eb-980f-9ff1da46c33c.png)

Each store that doesn't have wallet setup / hint is not dismissed now has warning icon:

![image](https://user-images.githubusercontent.com/5191402/96217361-0623bf00-0f48-11eb-8323-67d3cca8f9c4.png)

Hint also displayed on Rates page:

![image](https://user-images.githubusercontent.com/5191402/96217378-1176ea80-0f48-11eb-8b89-5d54feeae032.png)

I assume the biggest discussion will be on where we are persisting these hints. I've done first version with saving hints in StoreBlob. But could see this ending on client side... or maybe in UserPrefsCookie. Will see after additional discussion.

dstrukt commented:
Hell yeah!

Few thoughts: 

1. Personally, I don't think all of the explanation text underneath Wallet or Lightning is necessary. I think a clear tooltip explanation as to why you need to set up your derivation scheme to accept funds is all we need, until we update this section with a simple "Setup Wallet" CTA. 

If we really wanted to include more information it should be in the form of an expandable tooltip (or something along those lines) to help keep the focus clear. Ideally everything we're building is clear enough to navigate without a bunch of explanation, and we educate as we step through the process in bite-sized contextual pieces.

2. The wording for the wallet tooltips could a little clearer.

"A store requires a wallet to receive funds. Please provide a derivation scheme to set up your wallet."
&
"A connection to a lightning node is required to generate invoices and receive payments."

Perhaps?

3. As for the persistent, until complete, tooltip in the table rows, I imagine it would also be clickable and take you directly to the tab where you set up the wallet?

4. I still prefer the 

- General
- Wallet
- Lightning

list order.

I'm curious to see how this UX pattern feels when launched. I do like the simplicity of it... However, I still feel the CTA could be clearer and presented at the top of the page with a direct action vs. a series of warnings / hints. 

Maybe I'm wrong about this...shrug

Some visuals to correspond this long winded rant .. eventually the derivation scheme table would simplified with a singular CTA or a completed simplified component ... realizing I want to explore that..

🍻 

<img width="1277" alt="Screen Shot 2020-10-16 at 12 38 13 AM" src="https://user-images.githubusercontent.com/6250771/96226809-07ed8280-0f48-11eb-8312-6df04c04a37a.png">
<img width="1278" alt="Screen Shot 2020-10-16 at 12 37 27 AM" src="https://user-images.githubusercontent.com/6250771/96226807-058b2880-0f48-11eb-8f9c-3b57173ecd58.png">

rockstardev commented:
OK, we addressed all this feedback during Design call and it was decided to proceed with merging this as foundation, and then going for more changes after.

@NicolasDorier the biggest thing in here is you giving ACK to having Hints as part of Store Blob - see how I've implemented it. I do think this way should stick long term - this will give user ability to dismiss hints in case they don't want the hint sticking (for whatever reason). Different than driving hints from say `DerivationScheme` collection (where you don't have ability to permanently dismiss).

Alternative storage is to persist this in cookies, downside being that it won't work across browsers... it's really part of store. 
Also another consideration in case of cookies is that we would need to iterate cookie storage when displaying store listing.

However, no critical downside in any case... so checkout branch and let us know your feedback.

NicolasDorier commented on BTCPayServer/Models/StoreViewModels/StoreViewModel.cs:
```diff
@@ -90,6 +91,9 @@ public NetworkFeeMode NetworkFeeMode
         public bool PayJoinEnabled { get; set; }
 
+        public bool HintWallet { get; set; }
+        public bool HintLightning { get; set; }
```
I think just exposing `StoreHints` instead would prevent from maintaining the mapping. That said, not really important.

rockstardev replied:
I did it this way because view model wraps other properties from blob... this will work nicely if we add more hints to store and we don't want them on general page / in this view model.

NicolasDorier commented on BTCPayServer/Services/Stores/StoreRepository.cs:
```diff
@@ -182,6 +182,13 @@ public async Task CreateStore(string ownerId, StoreData storeData)
+            {
+                Wallet = true,
+                Lightning = true
+            };
```
would be better in ctor I would say, bu thats ok

rockstardev replied:
I wanted to leave default constructor empty because we need it from json deserialization and sometimes we want these props set to false, sometimes to true... found it easier like this.

NicolasDorier commented:
seems great, though I think you should highjack one of the selenium test to make sure the warning show up when the store is just created.

rockstardev commented:
OK, since @NicolasDorier signed off on this approach, I've tweaked the rest of the code - so hints are dismissed when wallet / lightning connection are set and added additional checks to Selenium test.

Merging it later today.

NicolasDorier commented on BTCPayServer.Tests/SeleniumTester.cs:
```diff
@@ -42,18 +42,19 @@ public async Task StartAsync()
+            if (Server.PayTester.InContainer)
+            {
+                // this must be first option https://stackoverflow.com/questions/53073411/selenium-webdriverexceptionchrome-failed-to-start-crashed-as-google-chrome-is#comment102570662_53073789
+                options.AddArgument("no-sandbox");
```
If this does not break the build, it is ok, but I think this mean some test might fail because chrome does not start test in a new profile.

pavlenex commented:
Hell yeaaaaah! 🚀 🚀 🚀
